monai/apps/reconstruction/ssim_loss.py

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from monai.utils.type_conversion import convert_to_tensor
+
+
+class SSIM(nn.Module):
+    """
+    Build a Pytorch version of the SSIM loss function based on the original formula of SSIM
+
+    Modified and adopted from:
+        https://github.com/facebookresearch/fastMRI/blob/main/banding_removal/fastmri/ssim_loss_mixin.py
+    """
+
+    def __init__(self, win_size: int = 7, k1: float = 0.01, k2: float = 0.03):
+        super().__init__()
+        self.win_size = win_size
+        self.k1, self.k2 = k1, k2
+        self.register_buffer("w", torch.ones(1, 1, win_size, win_size) / win_size**2)
+        np = win_size**2
+        self.cov_norm = np / (np - 1)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor, data_range: torch.Tensor) -> torch.Tensor:
+        """
+        y and x are of shape (B,C,H,W) where B is batch_size, C is number of channels, and (H,W) are height and width.
```

Comment:
Does it work for 3D images?

Reply:
Yes, **C** can also serve as the slice dimension to pass a volume to it. Will clarify that in the docstring.

Reply:
When C is used as slice dimension, that seems to be pseudo-3D, ssim is still computed on 2D patches, but averaged scross slices. What I was referring is true 3D SSIM